### PR TITLE
Fix error handling in Swarm dial() and listen_on()

### PR DESCRIPTION
**Pull request by tomaka**

### CODE REVIEW

**Comment by tomaka** [comment]
Rebased over master ; travis tests won't pass because of #61

File: libp2p-swarm/src/transport.rs [hunk 1]
```diff
@@ -579,14 +579,12 @@ impl<'a, T, C> UpgradedNode<T, C>
 					(n, <Bytes as PartialEq>::eq, t)
 				}
 				let iter = upgrade.protocol_names().map(iter_map);
-				let negotiated = multistream_select::listener_select_proto(connection, iter)
-					.map_err(|err| panic!("{:?}", err));      // TODO:
+				let negotiated = multistream_select::listener_select_proto(connection, iter);
 				negotiated.map(move |(upgrade_id, conn)| (upgrade_id, conn, upgrade, client_addr))
-					.map_err(|_| panic!())    // TODO:
+					.map_err(|err| IoError::new(IoErrorKind::Other, err))
```
Comment (Vurich) [inline libp2p-swarm/src/transport.rs]: Style nit: I prefer `use std::io;` to `use std::io::Error as IoError;`, but that could be argued as personal taste.

File: multistream-select/src/error.rs [hunk 2]
```diff
@@ -49,3 +51,36 @@ impl From<IoError> for ProtocolChoiceError {
...
+				"received a message from the remote that makes no sense in the current context"
+			},
+			ProtocolChoiceError::NoProtocolFound => {
+				"we don't support any protocol in common with the remote"
+			},
+		}
+	}
+
+	fn cause(&self) -> Option<&error::Error> {
+		match *self {
```
Comment (Vurich) [inline multistream-select/src/error.rs]: Style nit: maybe use an `if let`?

File: multistream-select/src/protocol/error.rs [hunk 3]
```diff
@@ -44,3 +46,39 @@ impl From<IoError> for MultistreamSelectError {
...
+			},
+			MultistreamSelectError::WrongProtocolName => {
+				"protocol names must always start with `/`, otherwise this error is returned"
+			},
+		}
+	}
+
+	fn cause(&self) -> Option<&error::Error> {
+		match *self {
+			MultistreamSelectError::IoError(ref err) => {
```
Comment (Vurich) [inline multistream-select/src/protocol/error.rs]: Why do we repeat `Error`/`MultistreamSelect` twice in this? Can it be `enum Error { Io(...), ... }`? We know we're in the multistream select crate already.

File: multistream-select/src/error.rs (hunk 2 above)
Comment (tomaka) [reply multistream-select/src/error.rs]: Nit: A match is better if we add more variants in the future

**Pull request merged**